File: app/src/main/java/com/n30dyn4m1c/photosphere/metadata/GPanoXmpInjector.java
package com.n30dyn4m1c.photosphere.metadata;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.EOFException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;

/**
 * Writes GPano XMP into a stitched equirectangular JPEG.
 *
 * <p>Without this the output is just a wide picture: what makes a viewer open a
 * file as a pannable 360° photo is an XMP packet in the {@code GPano} namespace, and
 * neither {@code ExifInterface} nor {@code Bitmap.compress} can put one there.
 *
 * <p><b>The pixels are never touched.</b> The JPEG is not decoded and not re-encoded —
 * this walks the file's marker segments, drops any XMP packet already present,
 * splices in a new {@code APP1} segment, and copies every remaining byte (including
 * the whole entropy-coded scan) through verbatim. Re-encoding to add metadata
 * would cost a generation of JPEG loss on an image that has already been through
 * one on the way in.
 *
 * <h3>Where the segment goes</h3>
 *
 * <p>A JPEG is {@code SOI} followed by marker segments; the metadata ones ({@code APPn}) sit at
 * the front, before the quantisation tables and the scan. The new packet is
 * placed after the last {@code APP0}/{@code APP1} — that is, after JFIF and after EXIF if
 * they are there — which is where readers expect to find it, and before anything
 * heavier such as an ICC profile.
 *
 * <p>Everything here is plain {@code java.io}, so it is unit-testable off-device.
 */
public final class GPanoXmpInjector {

    /** The Photo Sphere namespace itself. */
    public static final String GPANO_NAMESPACE = "http://ns.google.com/photos/1.0/panorama/";

    private static final String RDF_NAMESPACE = "http://www.w3.org/1999/02/22-rdf-syntax-ns#";

    /** Identifies an {@code APP1} segment as carrying XMP rather than EXIF. */
    private static final String XMP_NAMESPACE = "http://ns.adobe.com/xap/1.0/";

    /**
     * Second half of a packet too large for one segment. Never written here —
     * the GPano packet is a few hundred bytes — but recognised so that replacing
     * someone else's XMP cannot leave an orphaned continuation behind.
     */
    private static final String XMP_EXTENSION_NAMESPACE = "http://ns.adobe.com/xmp/extension/";

    /** Tool marker written into the packet, for anyone inspecting the file later. */
    private static final String TOOLKIT = "PhotoSphere";

    private static final byte[] XMP_SIGNATURE =
            (XMP_NAMESPACE + "\u0000").getBytes(StandardCharsets.US_ASCII);
    private static final byte[] XMP_EXTENSION_SIGNATURE =
            (XMP_EXTENSION_NAMESPACE + "\u0000").getBytes(StandardCharsets.US_ASCII);

    private static final int MARKER_PREFIX = 0xFF;
    private static final int MARKER_SOI = 0xD8;
    private static final int MARKER_APP0 = 0xE0;
    private static final int MARKER_APP1 = 0xE1;
    private static final int MARKER_APP15 = 0xEF;
    private static final int MARKER_COM = 0xFE;

    /** The two bytes of a segment's length field, which the field itself counts. */
    private static final int SEGMENT_LENGTH_BYTES = 2;

    /** Largest value that field can hold, and so the largest segment. */
    private static final int MAX_SEGMENT_LENGTH = 0xFFFF;

    /** Suffix of the file written beside the target while it is being rebuilt. */
    private static final String TEMP_SUFFIX = ".gpano.tmp";

    private GPanoXmpInjector() {
    }

    /**
     * Outcome of an in-place inject: the rewritten file on success, or the
     * failure that left the original untouched.
     */
    public static final class Result {
        private final File file;
        private final Throwable error;

        private Result(File file, Throwable error) {
            this.file = file;
            this.error = error;
        }

        public static Result success(File file) {
            return new Result(file, null);
        }

        public static Result failure(Throwable error) {
            return new Result(null, error);
        }

        public boolean isSuccess() {
            return error == null;
        }

        public boolean isFailure() {
            return error != null;
        }

        public File getFile() {
            return file;
        }

        public Throwable getError() {
            return error;
        }

        public File getOrThrow() throws IOException {
            if (error != null) {
                if (error instanceof IOException) {
                    throw (IOException) error;
                }
                throw new IOException(error.getMessage(), error);
            }
            return file;
        }
    }

    /**
     * Injects full-sphere GPano metadata into {@code file}, in place.
     *
     * <p>{@code imageWidth} and {@code imageHeight} must be the JPEG's real pixel dimensions;
     * they are what a viewer uses to map pixels onto the sphere, so a wrong value
     * shows up as a panorama that does not close or that is squashed at the
     * poles.
     *
     * <p>Blocking I/O over the whole file — call off the main thread. Failures come
     * back as a failed {@link Result} rather than an exception: a sphere whose metadata
     * could not be written is still a perfectly good image, and losing it over a
     * header would be the worse outcome.
     */
    public static Result inject(File file, int imageWidth, int imageHeight) {
        return inject(file, GPanoMetadata.forFullPano(imageWidth, imageHeight));
    }

    /** Injects {@code metadata} into {@code file}, in place. See the overload above. */
    public static Result inject(File file, GPanoMetadata metadata) {
        File temp = new File(file.getParentFile(), file.getName() + TEMP_SUFFIX);
        try {
            BufferedInputStream source = new BufferedInputStream(new FileInputStream(file));
            try {
                BufferedOutputStream sink = new BufferedOutputStream(new FileOutputStream(temp));
                try {
                    inject(source, sink, metadata);
                } finally {
                    sink.close();
                }
            } finally {
                source.close();
            }
            if (!temp.renameTo(file)) {
                // Some filesystems refuse to rename onto an existing name. The
                // original is fully readable until this point; replace it
                // atomically rather than delete-then-rename, which would open a
                // window where neither copy exists if the second call failed.
                try {
                    Files.move(temp.toPath(), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
                } catch (IOException e) {
                    throw new IOException(
                            "Could not move " + temp.getName() + " onto " + file.getName(), e);
                }
            }
            return Result.success(file);
        } catch (Exception e) {
            return Result.failure(e);
        } finally {
            if (temp.exists()) {
                temp.delete();
            }
        }
    }

    /**
     * Copies the JPEG on {@code source} to {@code sink} with {@code metadata} spliced into it.
     *
     * <p>Neither stream is closed. Throws {@link IOException} if {@code source} is not a JPEG,
     * ends mid-segment, or the packet will not fit in one {@code APP1} segment.
     */
    public static void inject(InputStream source, OutputStream sink, GPanoMetadata metadata)
            throws IOException {
        byte[] packet = buildXmpPacket(metadata).getBytes(StandardCharsets.UTF_8);
        int segmentLength = SEGMENT_LENGTH_BYTES + XMP_SIGNATURE.length + packet.length;
        if (segmentLength > MAX_SEGMENT_LENGTH) {
            throw new IOException(
                    "XMP packet is " + segmentLength + " bytes, too large for one segment");
        }

        if (source.read() != MARKER_PREFIX || source.read() != MARKER_SOI) {
            throw new IOException("Not a JPEG: the file does not start with SOI");
        }
        sink.write(MARKER_PREFIX);
        sink.write(MARKER_SOI);

        LeadingSegments leading = readLeadingSegments(source);

        // After JFIF and EXIF, before an ICC profile or anything else.
        int insertAt = 0;
        for (int i = 0; i < leading.segments.size(); i++) {
            int marker = leading.segments.get(i).marker;
            if (marker == MARKER_APP0 || marker == MARKER_APP1) {
                insertAt = i + 1;
            }
        }

        for (int i = 0; i < insertAt; i++) {
            leading.segments.get(i).writeTo(sink);
        }
        writeXmpSegment(sink, packet);
        for (int i = insertAt; i < leading.segments.size(); i++) {
            leading.segments.get(i).writeTo(sink);
        }

        // From the first non-metadata marker on, the file is copied byte for
        // byte: tables, frame header, and the entropy-coded scan untouched.
        sink.write(MARKER_PREFIX);
        sink.write(leading.nextMarker);
        copy(source, sink);
    }

    /**
     * The XMP packet, as the text that goes into the segment.
     *
     * <p>{@code UsePanoramaViewer} is the switch a viewer actually reads; the rest tells
     * it how to wrap the pixels around the sphere.
     */
    public static String buildXmpPacket(GPanoMetadata metadata) {
        StringBuilder out = new StringBuilder();
        out.append("<?xpacket begin=\"\uFEFF\" id=\"W5M0MpCehiHzreSzNTczkc9d\"?>\n");
        out.append("<x:xmpmeta xmlns:x=\"adobe:ns:meta/\" x:xmptk=\"").append(TOOLKIT).append("\">\n");
        out.append("  <rdf:RDF xmlns:rdf=\"").append(RDF_NAMESPACE).append("\">\n");
        out.append("    <rdf:Description rdf:about=\"\"\n");
        out.append("        xmlns:GPano=\"").append(GPANO_NAMESPACE).append("\"\n");
        out.append("        GPano:UsePanoramaViewer=\"")
                .append(asXmpBoolean(metadata.usePanoramaViewer)).append("\"\n");
        out.append("        GPano:ProjectionType=\"")
                .append(escapeXmlAttribute(metadata.projectionType)).append("\"\n");
        out.append("        GPano:FullPanoWidthPixels=\"")
                .append(metadata.fullPanoWidthPixels).append("\"\n");
        out.append("        GPano:FullPanoHeightPixels=\"")
                .append(metadata.fullPanoHeightPixels).append("\"\n");
        out.append("        GPano:CroppedAreaImageWidthPixels=")
                .append("\"").append(metadata.croppedAreaImageWidthPixels).append("\"\n");
        out.append("        GPano:CroppedAreaImageHeightPixels=")
                .append("\"").append(metadata.croppedAreaImageHeightPixels).append("\"\n");
        out.append("        GPano:CroppedAreaLeftPixels=\"")
                .append(metadata.croppedAreaLeftPixels).append("\"\n");
        out.append("        GPano:CroppedAreaTopPixels=\"")
                .append(metadata.croppedAreaTopPixels).append("\"/>\n");
        out.append("  </rdf:RDF>\n");
        out.append("</x:xmpmeta>\n");
        // "r" for read-only: no padding follows, so nothing may edit the packet
        // in place. Rewriting the segment, as this class does, is still fine.
        out.append("<?xpacket end=\"r\"?>");
        return out.toString();
    }

    /** One marker segment, held while the metadata block is reordered. */
    private static final class Segment {
        final int marker;
        final byte[] payload;

        Segment(int marker, byte[] payload) {
            this.marker = marker;
            this.payload = payload;
        }

        void writeTo(OutputStream sink) throws IOException {
            int length = SEGMENT_LENGTH_BYTES + payload.length;
            sink.write(MARKER_PREFIX);
            sink.write(marker);
            sink.write(length >>> 8);
            sink.write(length & 0xFF);
            sink.write(payload);
        }
    }

    private static final class LeadingSegments {
        final List<Segment> segments;
        final int nextMarker;

        LeadingSegments(List<Segment> segments, int nextMarker) {
            this.segments = segments;
            this.nextMarker = nextMarker;
        }
    }

    /**
     * Reads the run of metadata segments at the head of the file, dropping any
     * XMP already there, and stops on the first marker that is not one.
     *
     * <p>Returns those segments plus the marker it stopped on, which the caller
     * still has to write — it has been consumed from {@code source} but belongs to the
     * verbatim tail.
     */
    private static LeadingSegments readLeadingSegments(InputStream source) throws IOException {
        List<Segment> segments = new ArrayList<Segment>();
        while (true) {
            int prefix = source.read();
            if (prefix == -1) {
                throw new IOException("JPEG ended before any image data");
            }
            if (prefix != MARKER_PREFIX) {
                throw new IOException("Expected a marker, found 0x" + Integer.toHexString(prefix));
            }

            // Any number of 0xFF bytes may pad the gap before a marker.
            int marker = source.read();
            while (marker == MARKER_PREFIX) {
                marker = source.read();
            }
            if (marker == -1) {
                throw new IOException("JPEG ended on a bare marker prefix");
            }

            boolean isMetadata =
                    (marker >= MARKER_APP0 && marker <= MARKER_APP15) || marker == MARKER_COM;
            if (!isMetadata) {
                return new LeadingSegments(segments, marker);
            }

            int length = (readOrThrow(source) << 8) | readOrThrow(source);
            if (length < SEGMENT_LENGTH_BYTES) {
                throw new IOException(
                        "Segment 0x" + Integer.toHexString(marker) + " declares " + length + " bytes");
            }
            byte[] payload = readExactly(source, length - SEGMENT_LENGTH_BYTES);

            // Replacing rather than appending: two XMP packets in one file is
            // undefined, and readers generally take the first one they find.
            boolean isExistingXmp = marker == MARKER_APP1
                    && (startsWith(payload, XMP_SIGNATURE)
                    || startsWith(payload, XMP_EXTENSION_SIGNATURE));
            if (!isExistingXmp) {
                segments.add(new Segment(marker, payload));
            }
        }
    }

    private static void writeXmpSegment(OutputStream sink, byte[] packet) throws IOException {
        int length = SEGMENT_LENGTH_BYTES + XMP_SIGNATURE.length + packet.length;
        sink.write(MARKER_PREFIX);
        sink.write(MARKER_APP1);
        sink.write(length >>> 8);
        sink.write(length & 0xFF);
        sink.write(XMP_SIGNATURE);
        sink.write(packet);
    }

    /** XMP booleans are the words, capitalised — not {@code 1}/{@code 0} or lowercase. */
    private static String asXmpBoolean(boolean value) {
        return value ? "True" : "False";
    }

    private static String escapeXmlAttribute(String value) {
        StringBuilder out = new StringBuilder(value.length());
        for (int i = 0; i < value.length(); i++) {
            char character = value.charAt(i);
            switch (character) {
                case '&':
                    out.append("&amp;");
                    break;
                case '<':
                    out.append("&lt;");
                    break;
                case '>':
                    out.append("&gt;");
                    break;
                case '"':
                    out.append("&quot;");
                    break;
                case '\'':
                    out.append("&apos;");
                    break;
                default:
                    out.append(character);
                    break;
            }
        }
        return out.toString();
    }

    private static boolean startsWith(byte[] bytes, byte[] signature) {
        if (bytes.length < signature.length) {
            return false;
        }
        for (int i = 0; i < signature.length; i++) {
            if (bytes[i] != signature[i]) {
                return false;
            }
        }
        return true;
    }

    private static int readOrThrow(InputStream source) throws IOException {
        int value = source.read();
        if (value == -1) {
            throw new EOFException("JPEG ended inside a segment header");
        }
        return value;
    }

    private static byte[] readExactly(InputStream source, int count) throws IOException {
        byte[] bytes = new byte[count];
        int offset = 0;
        while (offset < count) {
            int read = source.read(bytes, offset, count - offset);
            if (read < 0) {
                throw new EOFException(
                        "JPEG ended " + offset + "/" + count + " bytes into a segment");
            }
            offset += read;
        }
        return bytes;
    }

    private static void copy(InputStream source, OutputStream sink) throws IOException {
        byte[] buffer = new byte[8192];
        int read;
        while ((read = source.read(buffer)) >= 0) {
            sink.write(buffer, 0, read);
        }
    }
}

File: app/src/main/java/com/n30dyn4m1c/photosphere/metadata/GPanoXmpInjector.kt
package com.n30dyn4m1c.photosphere.metadata

import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.EOFException
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import kotlin.math.roundToInt

/**
 * The Google Photo Sphere (GPano) properties that mark a JPEG as a 360° photo.
 *
 * Every length is in pixels of the image the metadata is attached to. The
 * `CroppedArea*` group describes which part of the full sphere the pixels
 * actually cover: for a full equirectangular frame it is the whole image, which
 * is what [forFullPano] produces. A partial capture keeps the full-pano
 * dimensions at the size the *complete* sphere would have been and places the
 * covered rectangle inside it — [forSphereRegion] expresses a ring capture that
 * way, so a viewer renders the band the frames actually cover and limits
 * panning to it instead of showing the black wedges where nothing was shot.
 */
data class GPanoMetadata(
    /** Width the complete sphere occupies. */
    val fullPanoWidthPixels: Int,
    /** Height the complete sphere occupies; half [fullPanoWidthPixels] at 2:1. */
    val fullPanoHeightPixels: Int,
    /** Width of the region of the sphere this image holds. */
    val croppedAreaImageWidthPixels: Int,
    /** Height of the region of the sphere this image holds. */
    val croppedAreaImageHeightPixels: Int,
    /** Left edge of that region within the full sphere. */
    val croppedAreaLeftPixels: Int,
    /** Top edge of that region within the full sphere. */
    val croppedAreaTopPixels: Int,
    /** Projection the pixels are in. Viewers only understand [PROJECTION_EQUIRECTANGULAR]. */
    val projectionType: String = PROJECTION_EQUIRECTANGULAR,
    /** Whether viewers should open this in a pannable sphere viewer. */
    val usePanoramaViewer: Boolean = true,
) {
    init {
        require(fullPanoWidthPixels > 0 && fullPanoHeightPixels > 0) {
            "full pano must have area, was ${fullPanoWidthPixels}x$fullPanoHeightPixels"
        }
        require(croppedAreaImageWidthPixels > 0 && croppedAreaImageHeightPixels > 0) {
            "cropped area must have area, was " +
                "${croppedAreaImageWidthPixels}x$croppedAreaImageHeightPixels"
        }
        require(croppedAreaLeftPixels >= 0 && croppedAreaTopPixels >= 0) {
            "cropped area origin must be inside the sphere, was " +
                "($croppedAreaLeftPixels, $croppedAreaTopPixels)"
        }
        require(croppedAreaLeftPixels + croppedAreaImageWidthPixels <= fullPanoWidthPixels) {
            "cropped area runs past the right edge of the sphere"
        }
        require(croppedAreaTopPixels + croppedAreaImageHeightPixels <= fullPanoHeightPixels) {
            "cropped area runs past the bottom edge of the sphere"
        }
        require(projectionType.isNotBlank()) { "projectionType must not be blank" }
    }

    companion object {
        /** The only projection Google Photos, Facebook and friends will render. */
        const val PROJECTION_EQUIRECTANGULAR = "equirectangular"

        /**
         * Metadata for an image that *is* the whole sphere: the cropped area
         * covers all of it, anchored at the origin.
         */
        fun forFullPano(imageWidth: Int, imageHeight: Int): GPanoMetadata = GPanoMetadata(
            fullPanoWidthPixels = imageWidth,
            fullPanoHeightPixels = imageHeight,
            croppedAreaImageWidthPixels = imageWidth,
            croppedAreaImageHeightPixels = imageHeight,
            croppedAreaLeftPixels = 0,
            croppedAreaTopPixels = 0,
        )

        /**
         * Metadata for an image that holds a *region* of the sphere, mapped
         * equirectangularly.
         *
         * The image covers [longitudeSpanDegrees] of longitude centred on
         * [centerLongitudeDegrees] and [latitudeSpanDegrees] of latitude centred
         * on [centerLatitudeDegrees]. A viewer maps the full-pano dimensions to
         * the whole 360°×180° sphere and renders the cropped area inside it, so
         * a ring capture — one band all the way around the horizon — is expressed
         * as the matching horizontal slice of a full sphere, and panning is
         * limited to the band the frames actually covered. A full sphere passed
         * through here resolves to the same values as [forFullPano].
         */
        fun forSphereRegion(
            imageWidth: Int,
            imageHeight: Int,
            longitudeSpanDegrees: Float,
            centerLongitudeDegrees: Float,
            latitudeSpanDegrees: Float,
            centerLatitudeDegrees: Float,
        ): GPanoMetadata {
            require(longitudeSpanDegrees in 1f..360f) { "longitude span: $longitudeSpanDegrees" }
            require(latitudeSpanDegrees in 1f..180f) { "latitude span: $latitudeSpanDegrees" }
            val fullWidth = (imageWidth * 360f / longitudeSpanDegrees).roundToInt()
            val fullHeight = (imageHeight * 180f / latitudeSpanDegrees).roundToInt()
            val left = ((centerLongitudeDegrees + 180f) / 360f * fullWidth).roundToInt() -
                imageWidth / 2
            val top = ((90f - centerLatitudeDegrees - latitudeSpanDegrees / 2f) / 180f * fullHeight)
                .roundToInt()
            return GPanoMetadata(
                fullPanoWidthPixels = fullWidth,
                fullPanoHeightPixels = fullHeight,
                croppedAreaImageWidthPixels = imageWidth,
                croppedAreaImageHeightPixels = imageHeight,
                croppedAreaLeftPixels = left.coerceAtLeast(0),
                croppedAreaTopPixels = top.coerceAtLeast(0),
            )
        }
    }
}

/**
 * Writes GPano XMP into a stitched equirectangular JPEG.
 *
 * Without this the output is just a wide picture: what makes a viewer open a
 * file as a pannable 360° photo is an XMP packet in the `GPano` namespace, and
 * neither `ExifInterface` nor `Bitmap.compress` can put one there.
 *
 * **The pixels are never touched.** The JPEG is not decoded and not re-encoded —
 * this walks the file's marker segments, drops any XMP packet already present,
 * splices in a new `APP1` segment, and copies every remaining byte (including
 * the whole entropy-coded scan) through verbatim. Re-encoding to add metadata
 * would cost a generation of JPEG loss on an image that has already been through
 * one on the way in.
 *
 * ### Where the segment goes
 *
 * A JPEG is `SOI` followed by marker segments; the metadata ones (`APPn`) sit at
 * the front, before the quantisation tables and the scan. The new packet is
 * placed after the last `APP0`/`APP1` — that is, after JFIF and after EXIF if
 * they are there — which is where readers expect to find it, and before anything
 * heavier such as an ICC profile.
 *
 * Everything here is plain `java.io`, so it is unit-testable off-device.
 */
object GPanoXmpInjector {

    /** The Photo Sphere namespace itself. */
    const val GPANO_NAMESPACE = "http://ns.google.com/photos/1.0/panorama/"

    private const val RDF_NAMESPACE = "http://www.w3.org/1999/02/22-rdf-syntax-ns#"

    /** Identifies an `APP1` segment as carrying XMP rather than EXIF. */
    private const val XMP_NAMESPACE = "http://ns.adobe.com/xap/1.0/"

    /**
     * Second half of a packet too large for one segment. Never written here —
     * the GPano packet is a few hundred bytes — but recognised so that replacing
     * someone else's XMP cannot leave an orphaned continuation behind.
     */
    private const val XMP_EXTENSION_NAMESPACE = "http://ns.adobe.com/xmp/extension/"

    /** Tool marker written into the packet, for anyone inspecting the file later. */
    private const val TOOLKIT = "PhotoSphere"

    private val XMP_SIGNATURE = "$XMP_NAMESPACE\u0000".toByteArray(Charsets.US_ASCII)
    private val XMP_EXTENSION_SIGNATURE =
        "$XMP_EXTENSION_NAMESPACE\u0000".toByteArray(Charsets.US_ASCII)

    private const val MARKER_PREFIX = 0xFF
    private const val MARKER_SOI = 0xD8
    private const val MARKER_APP0 = 0xE0
    private const val MARKER_APP1 = 0xE1
    private const val MARKER_APP15 = 0xEF
    private const val MARKER_COM = 0xFE

    /** The two bytes of a segment's length field, which the field itself counts. */
    private const val SEGMENT_LENGTH_BYTES = 2

    /** Largest value that field can hold, and so the largest segment. */
    private const val MAX_SEGMENT_LENGTH = 0xFFFF

    /** Suffix of the file written beside the target while it is being rebuilt. */
    private const val TEMP_SUFFIX = ".gpano.tmp"

    /**
     * Injects full-sphere GPano metadata into [file], in place.
     *
     * [imageWidth] and [imageHeight] must be the JPEG's real pixel dimensions;
     * they are what a viewer uses to map pixels onto the sphere, so a wrong value
     * shows up as a panorama that does not close or that is squashed at the
     * poles.
     *
     * Blocking I/O over the whole file — call off the main thread. Failures come
     * back as a failed [Result] rather than an exception: a sphere whose metadata
     * could not be written is still a perfectly good image, and losing it over a
     * header would be the worse outcome.
     */
    fun inject(file: File, imageWidth: Int, imageHeight: Int): Result<File> =
        inject(file, GPanoMetadata.forFullPano(imageWidth, imageHeight))

    /** Injects [metadata] into [file], in place. See the overload above. */
    fun inject(file: File, metadata: GPanoMetadata): Result<File> = runCatching {
        val temp = file.resolveSibling(file.name + TEMP_SUFFIX)
        try {
            BufferedInputStream(FileInputStream(file)).use { source ->
                BufferedOutputStream(FileOutputStream(temp)).use { sink ->
                    inject(source, sink, metadata)
                }
            }
            if (!temp.renameTo(file)) {
                // Some filesystems refuse to rename onto an existing name. The
                // original is fully readable until this point; replace it
                // atomically rather than delete-then-rename, which would open a
                // window where neither copy exists if the second call failed.
                try {
                    Files.move(temp.toPath(), file.toPath(), StandardCopyOption.REPLACE_EXISTING)
                } catch (e: IOException) {
                    throw IOException("Could not move ${temp.name} onto ${file.name}", e)
                }
            }
            file
        } finally {
            if (temp.exists()) temp.delete()
        }
    }

    /**
     * Copies the JPEG on [source] to [sink] with [metadata] spliced into it.
     *
     * Neither stream is closed. Throws [IOException] if [source] is not a JPEG,
     * ends mid-segment, or the packet will not fit in one `APP1` segment.
     */
    @Throws(IOException::class)
    fun inject(source: InputStream, sink: OutputStream, metadata: GPanoMetadata) {
        val packet = buildXmpPacket(metadata).toByteArray(Charsets.UTF_8)
        val segmentLength = SEGMENT_LENGTH_BYTES + XMP_SIGNATURE.size + packet.size
        if (segmentLength > MAX_SEGMENT_LENGTH) {
            throw IOException("XMP packet is $segmentLength bytes, too large for one segment")
        }

        if (source.read() != MARKER_PREFIX || source.read() != MARKER_SOI) {
            throw IOException("Not a JPEG: the file does not start with SOI")
        }
        sink.write(MARKER_PREFIX)
        sink.write(MARKER_SOI)

        val (leading, nextMarker) = readLeadingSegments(source)

        // After JFIF and EXIF, before an ICC profile or anything else.
        val insertAt = leading.indexOfLast {
            it.marker == MARKER_APP0 || it.marker == MARKER_APP1
        } + 1

        leading.take(insertAt).forEach { it.writeTo(sink) }
        writeXmpSegment(sink, packet)
        leading.drop(insertAt).forEach { it.writeTo(sink) }

        // From the first non-metadata marker on, the file is copied byte for
        // byte: tables, frame header, and the entropy-coded scan untouched.
        sink.write(MARKER_PREFIX)
        sink.write(nextMarker)
        source.copyTo(sink)
    }

    /**
     * The XMP packet, as the text that goes into the segment.
     *
     * `UsePanoramaViewer` is the switch a viewer actually reads; the rest tells
     * it how to wrap the pixels around the sphere.
     */
    fun buildXmpPacket(metadata: GPanoMetadata): String = buildString {
        append("<?xpacket begin=\"\uFEFF\" id=\"W5M0MpCehiHzreSzNTczkc9d\"?>\n")
        append("<x:xmpmeta xmlns:x=\"adobe:ns:meta/\" x:xmptk=\"$TOOLKIT\">\n")
        append("  <rdf:RDF xmlns:rdf=\"$RDF_NAMESPACE\">\n")
        append("    <rdf:Description rdf:about=\"\"\n")
        append("        xmlns:GPano=\"$GPANO_NAMESPACE\"\n")
        append("        GPano:UsePanoramaViewer=\"${metadata.usePanoramaViewer.asXmpBoolean()}\"\n")
        append("        GPano:ProjectionType=\"${metadata.projectionType.escapeXmlAttribute()}\"\n")
        append("        GPano:FullPanoWidthPixels=\"${metadata.fullPanoWidthPixels}\"\n")
        append("        GPano:FullPanoHeightPixels=\"${metadata.fullPanoHeightPixels}\"\n")
        append(
            "        GPano:CroppedAreaImageWidthPixels=" +
                "\"${metadata.croppedAreaImageWidthPixels}\"\n"
        )
        append(
            "        GPano:CroppedAreaImageHeightPixels=" +
                "\"${metadata.croppedAreaImageHeightPixels}\"\n"
        )
        append("        GPano:CroppedAreaLeftPixels=\"${metadata.croppedAreaLeftPixels}\"\n")
        append("        GPano:CroppedAreaTopPixels=\"${metadata.croppedAreaTopPixels}\"/>\n")
        append("  </rdf:RDF>\n")
        append("</x:xmpmeta>\n")
        // "r" for read-only: no padding follows, so nothing may edit the packet
        // in place. Rewriting the segment, as this class does, is still fine.
        append("<?xpacket end=\"r\"?>")
    }

    /** One marker segment, held while the metadata block is reordered. */
    private class Segment(val marker: Int, val payload: ByteArray) {
        fun writeTo(sink: OutputStream) {
            val length = SEGMENT_LENGTH_BYTES + payload.size
            sink.write(MARKER_PREFIX)
            sink.write(marker)
            sink.write(length ushr 8)
            sink.write(length and 0xFF)
            sink.write(payload)
        }
    }

    /**
     * Reads the run of metadata segments at the head of the file, dropping any
     * XMP already there, and stops on the first marker that is not one.
     *
     * Returns those segments plus the marker it stopped on, which the caller
     * still has to write — it has been consumed from [source] but belongs to the
     * verbatim tail.
     */
    private fun readLeadingSegments(source: InputStream): Pair<List<Segment>, Int> {
        val segments = ArrayList<Segment>()
        while (true) {
            val prefix = source.read()
            if (prefix == -1) throw IOException("JPEG ended before any image data")
            if (prefix != MARKER_PREFIX) {
                throw IOException("Expected a marker, found 0x${prefix.toString(16)}")
            }

            // Any number of 0xFF bytes may pad the gap before a marker.
            var marker = source.read()
            while (marker == MARKER_PREFIX) marker = source.read()
            if (marker == -1) throw IOException("JPEG ended on a bare marker prefix")

            val isMetadata = marker in MARKER_APP0..MARKER_APP15 || marker == MARKER_COM
            if (!isMetadata) return segments to marker

            val length = (source.readOrThrow() shl 8) or source.readOrThrow()
            if (length < SEGMENT_LENGTH_BYTES) {
                throw IOException("Segment 0x${marker.toString(16)} declares $length bytes")
            }
            val payload = source.readExactly(length - SEGMENT_LENGTH_BYTES)

            // Replacing rather than appending: two XMP packets in one file is
            // undefined, and readers generally take the first one they find.
            val isExistingXmp = marker == MARKER_APP1 &&
                (payload.startsWith(XMP_SIGNATURE) || payload.startsWith(XMP_EXTENSION_SIGNATURE))
            if (!isExistingXmp) segments += Segment(marker, payload)
        }
    }

    private fun writeXmpSegment(sink: OutputStream, packet: ByteArray) {
        val length = SEGMENT_LENGTH_BYTES + XMP_SIGNATURE.size + packet.size
        sink.write(MARKER_PREFIX)
        sink.write(MARKER_APP1)
        sink.write(length ushr 8)
        sink.write(length and 0xFF)
        sink.write(XMP_SIGNATURE)
        sink.write(packet)
    }

    /** XMP booleans are the words, capitalised — not `1`/`0` or lowercase. */
    private fun Boolean.asXmpBoolean(): String = if (this) "True" else "False"

    private fun String.escapeXmlAttribute(): String = buildString(length) {
        this@escapeXmlAttribute.forEach { character ->
            when (character) {
                '&' -> append("&amp;")
                '<' -> append("&lt;")
                '>' -> append("&gt;")
                '"' -> append("&quot;")
                '\'' -> append("&apos;")
                else -> append(character)
            }
        }
    }

    private fun ByteArray.startsWith(signature: ByteArray): Boolean =
        size >= signature.size && signature.indices.all { this[it] == signature[it] }

    private fun InputStream.readOrThrow(): Int =
        read().also { if (it == -1) throw EOFException("JPEG ended inside a segment header") }

    private fun InputStream.readExactly(count: Int): ByteArray {
        val bytes = ByteArray(count)
        var offset = 0
        while (offset < count) {
            val read = read(bytes, offset, count - offset)
            if (read < 0) throw EOFException("JPEG ended $offset/$count bytes into a segment")
            offset += read
        }
        return bytes
    }
}
